transformer_engine/pytorch/csrc/extensions/permutation.cu

```diff
@@ -0,0 +1,164 @@
+
+    workspace.push_back(sorted_indices);
+    workspace.push_back(row_id);
+    workspace.push_back(sorted_row_id);
+    workspace.push_back(temp_storage);
+  }
+
+  int *indices_ptr = reinterpret_cast<int *>(getDataPtr(indices, 0));
+  int *sorted_indices_ptr = reinterpret_cast<int *>(getDataPtr(workspace[0], 0));
+  int *row_id_ptr = reinterpret_cast<int *>(getDataPtr(workspace[1], 0));
+  int *sorted_row_id_ptr = reinterpret_cast<int *>(getDataPtr(workspace[2], 0));
+
+  void *d_temp_storage = getDataPtr(workspace[3], 0);
+  size_t temp_storage_bytes = std::numeric_limits<size_t>::max();
+
+  cub::DeviceRadixSort::SortPairs(d_temp_storage, temp_storage_bytes, indices_ptr,
+                                  sorted_indices_ptr, row_id_ptr, sorted_row_id_ptr,
+                                  num_tokens * topK);
+
+  // Activations type
+  at::ScalarType _st;
+  if (dtype == transformer_engine::DType::kFloat8E4M3 ||
+      dtype == transformer_engine::DType::kFloat8E5M2)
+    _st = at::ScalarType::Float;
+  else
+    _st = input.scalar_type();
+
+  // Output buffer alloc
+  num_out_tokens = (num_out_tokens > 0) ? num_out_tokens : num_tokens * topK;
+  Tensor permuted_output = torch::empty(
+      {num_out_tokens, num_cols}, torch::dtype(_st).device(torch::kCUDA).requires_grad(false));
+  Tensor row_id_map = torch::empty(
+      {num_tokens * topK}, torch::dtype(torch::kInt32).device(torch::kCUDA).requires_grad(false));
+
+  int *row_id_map_ptr = reinterpret_cast<int *>(getDataPtr(row_id_map, 0));
+  auto stream = at::cuda::getCurrentCUDAStream().stream();
+
+  void *input_ptr = getDataPtr(input, 0);
+  void *permuted_output_ptr = getDataPtr(permuted_output, 0);
+
+  if (dtype == transformer_engine::DType::kFloat8E4M3 ||
+      dtype == transformer_engine::DType::kFloat8E5M2)
+    num_cols *= 4;
+
+  nvte_permute(input_ptr, permuted_output_ptr, dtype, sorted_row_id_ptr, row_id_map_ptr, nullptr,
+               num_tokens, topK, num_cols, num_out_tokens, nullptr, nullptr, stream);
+
+  return std::make_tuple(permuted_output, row_id_map, workspace);
+}
+
+Tensor moe_permute_bwd(Tensor input, const transformer_engine::DType dtype, Tensor row_id_map,
+                       Tensor prob, int64_t num_tokens, int64_t topK) {
+  return moe_unpermute_fwd(input, dtype, row_id_map, prob, num_tokens, topK);
+}
+
+Tensor moe_unpermute_fwd(Tensor input, const transformer_engine::DType dtype, Tensor row_id_map,
+                         Tensor prob, int64_t num_tokens, int64_t topK) {
+  int num_cols = input.size(1);
+
+  // Activations type
+  at::ScalarType _st;
+  if (dtype == transformer_engine::DType::kFloat8E4M3 ||
+      dtype == transformer_engine::DType::kFloat8E5M2)
+    _st = at::ScalarType::Float;
+  else
+    _st = input.scalar_type();
+
+  // Output buffer alloc
+  Tensor unpermuted_output = torch::empty(
+      {num_tokens, num_cols}, torch::dtype(_st).device(torch::kCUDA).requires_grad(false));
```

Comment:
Hi, 
TE uses `at::Tensor` and has a different way to create an empty tensor.  Please refer to this 
https://github.com/NVIDIA/TransformerEngine/blob/0b303dad4c66693351dda93e11c7b61b873751b0/transformer_engine/pytorch/csrc/extensions/cast.cu#L43-L58

Reply:
It seems like TE uses `torch::empty` as well.
https://github.com/NVIDIA/TransformerEngine/blob/9c127ef5b1ee10d466acf633f4a0ad3c8914cf1b/transformer_engine/pytorch/csrc/extensions/apply_rope.cu#L46